File: test/scripts/01-context-switching/correct.py
#!/usr/bin/env python3
"""
CORRECT: Limit threads to CPU cores
Resume: "Reduce Context Switching" - Match thread count to cores for CPU-bound work.

Run: DURATION_SEC=60 python correct.py
Measure: vmstat 1, pidstat -w -> fewer context switches
"""

import os
import signal
import threading
import multiprocessing
import time

ITERATIONS = 100_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False


def cpu_busy_work():
    total = 0
    while not stop:
        for i in range(ITERATIONS):
            total += i * (i + 1)
    return total


def main():
    global stop
    num_cores = multiprocessing.cpu_count()
    num_threads = num_cores  # Match threads to cores
    print(f"CORRECT: {num_threads} threads (= {num_cores} cores)")
    print(f"Running {DURATION_SEC}s. Measure: vmstat 1, pidstat -w")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))

    threads = [threading.Thread(target=cpu_busy_work) for _ in range(num_threads)]
    for t in threads:
        t.start()
    time.sleep(DURATION_SEC)
    stop = True
    for t in threads:
        t.join(timeout=2)
    print("Stopped")


if __name__ == "__main__":
    main()

File: test/scripts/01-context-switching/correct.js
#!/usr/bin/env node
/**
 * CORRECT: Limit workers to CPU cores
 * Run: DURATION_SEC=60 node correct.js
 */
const { Worker, isMainThread, workerData } = require('worker_threads');
const os = require('os');

const numWorkers = Math.max(1, (os.cpus().length || 4));
const ITERATIONS = 1_000_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);

if (isMainThread) {
  const sab = new SharedArrayBuffer(4);
  const stopFlag = new Int32Array(sab);
  console.log('CORRECT:', numWorkers, 'workers (= cores). Running', DURATION_SEC, 's');
  const workers = [];
  for (let i = 0; i < numWorkers; i++) {
    workers.push(new Promise((r) => {
      const w = new Worker(__filename, { workerData: { sab } });
      w.on('exit', r);
    }));
  }
  setTimeout(() => { Atomics.store(stopFlag, 0, 1); Promise.all(workers).then(() => console.log('Stopped')); }, DURATION_SEC * 1000);
  process.on('SIGINT', () => Atomics.store(stopFlag, 0, 1));
} else {
  const { sab } = workerData;
  const stopFlag = new Int32Array(sab);
  while (!Atomics.load(stopFlag, 0)) {
    let total = 0;
    for (let i = 0; i < ITERATIONS; i++) total += i * (i + 1);
  }
}
